packages/react-devtools-shared/src/dynamicImportCache.js

```diff
@@ -0,0 +1,159 @@
+// Otherwise, refreshing the inspected element cache would also clear this cache.
+// Modules are static anyway.
+const moduleLoaderFunctionToModuleMap: Map<
+  ModuleLoaderFunction,
+  Module,
+> = new Map();
+
+function readRecord<T>(record: Record<T>): ResolvedRecord<T> | RejectedRecord {
+  if (record.status === Resolved) {
+    // This is just a type refinement.
+    return record;
+  } else if (record.status === Rejected) {
+    // This is just a type refinement.
+    return record;
+  } else {
+    throw record.value;
+  }
+}
+
+// TODO Flow type
+export function loadModule(moduleLoaderFunction: ModuleLoaderFunction): Module {
+  let record = moduleLoaderFunctionToModuleMap.get(moduleLoaderFunction);
+
+  if (__DEBUG__) {
+    console.log(
+      `[dynamicImportCache] loadModule("${moduleLoaderFunction.name}")`,
+    );
+  }
+
+  if (!record) {
+    const callbacks = new Set();
+    const wakeable: Wakeable = {
+      then(callback) {
+        callbacks.add(callback);
+      },
+    };
+
+    const wake = () => {
+      if (timeoutID) {
+        clearTimeout(timeoutID);
+        timeoutID = null;
+      }
+
+      // This assumes they won't throw.
+      callbacks.forEach(callback => callback());
+      callbacks.clear();
+    };
+
+    const newRecord: Record<Module> = (record = {
+      status: Pending,
+      value: wakeable,
+    });
+
+    let didTimeout = false;
+
+    moduleLoaderFunction().then(
+      module => {
+        if (__DEBUG__) {
+          console.log(
+            `[dynamicImportCache] loadModule("${moduleLoaderFunction.name}") then()`,
+          );
+        }
+
+        if (didTimeout) {
+          return;
+        }
+
+        const resolvedRecord = ((newRecord: any): ResolvedRecord<Module>);
+        resolvedRecord.status = Resolved;
+        resolvedRecord.value = module;
+
+        wake();
+      },
+      error => {
+        if (__DEBUG__) {
+          console.log(
+            `[dynamicImportCache] loadModule("${moduleLoaderFunction.name}") catch()`,
+          );
+        }
+
+        if (didTimeout) {
+          return;
+        }
+
+        console.log(error);
+
+        const thrownRecord = ((newRecord: any): RejectedRecord);
+        thrownRecord.status = Rejected;
+        thrownRecord.value = null;
+
+        wake();
+      },
+    );
+
+    // Eventually timeout and stop trying to load the module.
+    let timeoutID = setTimeout(function onTimeout() {
+      if (__DEBUG__) {
+        console.log(
+          `[dynamicImportCache] loadModule("${moduleLoaderFunction.name}") onTimeout()`,
+        );
+      }
+
+      timeoutID = null;
+
+      didTimeout = true;
+
+      const timedoutRecord = ((newRecord: any): RejectedRecord);
+      timedoutRecord.status = Rejected;
+      timedoutRecord.value = null;
+
+      wake();
+    }, TIMEOUT);
+
+    moduleLoaderFunctionToModuleMap.set(moduleLoaderFunction, record);
+  }
+
+  const response = readRecord(record).value;
+  return response;
```

Comment:
cool! btw i know you've mentioned we use suspense caches for other things in dev tools. would it make sense to reuse any of those? or is it easier to write this one from scratch?

Reply:
They serve different purposes (have slightly different implementation, Flow typing, need to be purged/reset independently) so I think it makes sense to keep them separate 😄